Fix duplicates in AdjustableAudioComponent's adjustments

MrTheMake opened the pull request:
Resolves https://github.com/ppy/osu/issues/819

Supersedes https://github.com/ppy/osu-framework/pull/826

Jorolf commented on osu.Framework/Audio/AdjustableAudioComponent.cs:
```diff
@@ -78,34 +78,46 @@ public void AddAdjustmentDependency(AdjustableAudioComponent component)
                     frequencyAdjustments.Add(adjustBindable);
                     break;
                 case AdjustableProperty.Volume:
+                    if (volumeAdjustments.Contains(adjustBindable)) return;
```
use a hashset instead of a list...?

Jorolf replied:
https://msdn.microsoft.com/en-us/library/bb359438(v=vs.110).aspx#Anchor_7
"The HashSet<T> class is based on the model of mathematical sets and provides high-performance set operations similar to accessing the keys of the Dictionary<TKey, TValue> or Hashtable collections. In simple terms, the HashSet<T> class can be thought of as a Dictionary<TKey, TValue> collection without values.

A HashSet<T> collection is not sorted and cannot contain duplicate elements. If order or element duplication is more important than performance for your application, consider using the List<T> class together with the Sort method."

MrTheMake replied:
Used this class now. Didn't know C# provides such a class.

huoyaoyuan replied:
`HashSet` is designed for better `Contains` operation, while `List` is more array-like.

Jorolf replied:
Me neither until I googled it ;)

turbedi commented on osu.Framework/Audio/AdjustableAudioComponent.cs:
```diff
@@ -101,11 +101,20 @@ public void AddAdjustment(AdjustableProperty type, BindableDouble adjustBindable
+                    balanceAdjustments.Remove(adjustBindable);
+                    break;
+                case AdjustableProperty.Frequency:
+                    volumeAdjustments.Remove(adjustBindable);
```
`frequencyAdjustments`, not `volumeAdjustments`

turbedi commented on osu.Framework/Audio/AdjustableAudioComponent.cs:
```diff
@@ -101,11 +101,20 @@ public void AddAdjustment(AdjustableProperty type, BindableDouble adjustBindable
+                    volumeAdjustments.Remove(adjustBindable);
+                    break;
+                case AdjustableProperty.Volume:
+                    frequencyAdjustments.Remove(adjustBindable);
```
`volumeAdjustments`, not `frequencyAdjustments`

peppy commented:
Did you find out why components are being added twice in the first place?

MrTheMake commented:
As discussed in discord, I've added an exception when attempting to add an already added dependency. Moreover, the only case where this is attempted is also fixed.
